### DISCUSSION TITLE
Preparation for version 1

### DISCUSSION THREAD

**Discussion by lorinkoz** [lorinkoz/django-pgschemas#161] [General]
Hi everyone! I am making preparations to _finally_ release version 1 of `pgschemas` and formally mark it as "production stable". I have a few topics in which I could use your input, as consumer of the library. Thanks in advance!

## `--parallel` executor

All django commands have a `--parallel` option that runs the command on the specified schemas using multi-threading. This used to be a big thing when I first created this library. I no longer think that running a management command efficiently across all tenants should be a responsibility of this package. I am thinking of deprecating / removing this option, and suggest consumers to use an external task queue in order to obtain the same behavior. What do you think?

## `contrib.channels2` and `contrib.channels3`

It's quite hard for me to keep up with all versions of Django, let alone of channels. I am thinking of removing the channels contrib and make the code available as an example, for every consumer of the package to implement locally. What do you think?

**Comment by merabtenei** [comment]
Hi @lorinkoz, thanks for maintaining this library, I went with django-pgschemas over [django-tenants](https://github.com/django-tenants/django-tenants) because of those contrib.channels. For me this is something that was missing on django-tenants.
Keeping support for channels would be great and much appreciated.

**Reply by tomturner** [reply]
Hi @lorinkoz I always keep an eye on project! 😄    I know the feeling keeping up with django and channel.  I use [django-tenants](https://github.com/django-tenants/django-tenants) and django channel for my project. Perhaps we could create a helper project that allows people to use django channels and with either this project or [django-tenants](https://github.com/django-tenants/django-tenants).  Just a idea.

**Reply by merabtenei** [reply]
@tomturner that's a great idea, I used django-tenants on a previous project. We didn't find any tips on using channels with django-tenants.

**Reply by lorinkoz** [reply]
@tomturner Maybe having an external package is not necessary after all. It could even be burdensome to maintain, considering all the variations of `Django` and `channels` and `tenants` and `pgschemas`. But maybe there's an easier path...

The only two things that `pgschemas` provides regarding `channels` are:

- https://github.com/lorinkoz/django-pgschemas/blob/master/django_pgschemas/contrib/channels3/auth.py
- https://github.com/lorinkoz/django-pgschemas/blob/master/django_pgschemas/contrib/channels3/router.py

They are really nothing more than a copy of `channels` auth and router, only adding the bits so that it works with the active tenant.

Because this seems to have been the major cause for @merabtenei to stop using `django-tenants`, I wonder if you could provide a similar layer on `tenants`.

The main reason I was considering removing this from this package is because it's been a while since I've been actively working in a project that uses either `tenants` or `pgschemas` and I don't have a feeling anymore of how things are in that regard. But because you have it more fresh in your daily routine, maybe it won't be difficult to add the missing bits to `tenants`. What do you think?

I almost forget something important. I never got to properly test the channels code. I basically settled with making it very clear here https://django-pgschemas.readthedocs.io/en/stable/contrib.html#channels-websockets. I think I could afford it because `pgschemas` was not officially production-ready, but I don't know to what extent you could do the same in `tenants`. Maybe you have already comprehensive tests in your code? That would be awesome.

**Reply by lorinkoz** [reply]
Ah, I misread. You wrote "helper project" and not "helper package". Sorry for the confusion. That does make a lot of sense. However what I said remains, if you wanted to include some foundation to use `channels` with `tenants`.

**Reply by merabtenei** [reply]
We are a tiny team and we would like to focus on the main part of our projects, so we prefer to use anything that has something ready to go in this case the ```contrib.channels```. That's what motivated our move to pgschemas.

**Comment by tomturner** [comment]
I did mean helper package. I might create a help file on how to set channel up as it really easy.

**Comment by condedynamics** [comment]
How about adding a schema_context utility just as in the django-tenants?

something to make a job like this possible and easy because this doesn't even work...

    def get(self, request, format=None):
        """
        Return the singleton object of the Config.
        """
        schema_name = connection.schema_name
        with TenantModel.objects.get(schema_name=schema_name):
            # get_solo will create the item if it does not already exist
            config = Config.get_solo()
            serializer = ConfigReadWriteSerializer(config)
            return Response(serializer.data)

**Reply by lorinkoz** [reply]
Hmm, this should be working already.

**Reply by condedynamics** [reply]
Yes it worked when i found this out and used it:

from django_pgschemas.schema import get_current_schema
schema_name = get_current_schema().schema_name

instead this never did work for me:

from django.db import connection
schema_name = connection.schema_name

**Reply by condedynamics** [reply]
But this is what I was suggesting

from django_tenants.utils import schema_context

with schema_context(schema_name):
    # All commands here are ran under the schema `schema_name`

# Restores the `SEARCH_PATH` to its original value

**Reply by ktowen** [reply]
You can use the SchemaDescriptor like this 

```python3
from django_pgschemas.schema import get_current_schema, SchemaDescriptor

def current_schema_name():
    return get_current_schema().schema_name

schema_name = "name"
with SchemaDescriptor.create(schema_name):
    print(current_schema_name())
print(current_schema_name())
```

**Comment by lorinkoz** [comment]
Thank you all!
